### PR TITLE
Align Storage Access enums to spec

### LINKED ISSUE #6599: Strange usage of TextureFormatFeatureFlags::STORAGE_READ_WRITE
idk whats going on here but it looks wrong. if its not wrong it needs docs. why is read_write needed for read_only? why is write_only never checked?

https://github.com/gfx-rs/wgpu/blob/c22c062b54857e0598c0cccfb4d589406e1393b0/wgpu-core/src/device/resource.rs#L2496-L2506

### PR DESCRIPTION
**Pull request by atlv24**
**Connections**
Fixes #6599 

**Description**
Rename enum variants and change logic to match webgpu storage access spec as concluded in #6599.
Write-only storage access is now verified correctly, and as such two tests had to be updated to use TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES to get write-only access on the formats they were using.

**Future work**
- [x] make RON usages enum string based instead of hardcoded integers #6644
- [ ] renumber bitfields to not be all jumbled
- [ ] Give `BufferUses::STORAGE_WRITE_ONLY` and `TextureUses::STORAGE_WRITE_ONLY` a different bit?
- [ ] evaluate whether mapping `wgt::BufferUsages::STORAGE` -> `hal::BufferUses::STORAGE_READ_WRITE` and `wgt::TextureUsages::STORAGE_BINDING` -> `hal::TextureUses::STORAGE_READ_WRITE` in `wgpu-core/src/conv.rs` is correct.
- [ ] ~~Figure out what the hell is going on with `BufferBarrier`/`TextureBarrier` `usage: Range<BufferUses>` and `usage: Range<TextureUses>`. This smells real fishy, ranges of bitfields dont sound like they should work. They're not ordered.~~ #6645

**Testing**
Existing tests pass.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

File: wgpu-core/src/device/resource.rs [hunk 1]
```diff
@@ -2489,24 +2499,36 @@ impl Device {
                 }
 
                 let internal_use = match access {
-                    wgt::StorageTextureAccess::WriteOnly => hal::TextureUses::STORAGE_READ_WRITE,
+                    wgt::StorageTextureAccess::WriteOnly => {
+                        if !view.format_features.flags.intersects(
+                            wgt::TextureFormatFeatureFlags::STORAGE_WRITE_ONLY
+                                | wgt::TextureFormatFeatureFlags::STORAGE_READ_WRITE,
+                        ) {
+                            println!(
```
Comment (nical) [inline wgpu-core/src/device/resource.rs]: Let's remove the println. If the error is not descriptive enough we can look into improving it instead.

File: wgpu-core/src/device/resource.rs [hunk 2]
```diff
@@ -2489,24 +2499,36 @@ impl Device {
                 }
 
                 let internal_use = match access {
-                    wgt::StorageTextureAccess::WriteOnly => hal::TextureUses::STORAGE_READ_WRITE,
+                    wgt::StorageTextureAccess::WriteOnly => {
+                        if !view.format_features.flags.intersects(
+                            wgt::TextureFormatFeatureFlags::STORAGE_WRITE_ONLY
+                                | wgt::TextureFormatFeatureFlags::STORAGE_READ_WRITE,
```
Comment (nical) [inline wgpu-core/src/device/resource.rs:2508]: Textures that support write-only all support read+write, so we don't need to check for read+write here, unless I'm missing something.

File: wgpu-core/src/device/resource.rs [hunk 3]
```diff
@@ -2489,24 +2499,36 @@ impl Device {
...
+                    }
                     wgt::StorageTextureAccess::ReadOnly => {
-                        if !view
-                            .format_features
-                            .flags
-                            .contains(wgt::TextureFormatFeatureFlags::STORAGE_WRITE)
-                        {
+                        if !view.format_features.flags.intersects(
+                            wgt::TextureFormatFeatureFlags::STORAGE_READ_ONLY
+                                | wgt::TextureFormatFeatureFlags::STORAGE_READ_WRITE,
```
Comment (nical) [inline wgpu-core/src/device/resource.rs:2517]: Same thing, I don't think we need to check for read+write capability here.

File: wgpu-hal/src/lib.rs [hunk 4]
```diff
@@ -1543,8 +1543,10 @@ bitflags!(
         /// Format can be sampled with a min/max reduction sampler.
         const SAMPLED_MINMAX = 1 << 2;
 
+        /// Format can be used as storage with read-only access.
+        const STORAGE_READ_ONLY = 1 << 16;
         /// Format can be used as storage with write-only access.
-        const STORAGE_WRITE = 1 << 3;
+        const STORAGE_WRITE_ONLY = 1 << 3;
         /// Format can be used as storage with read and read/write access.
```
Comment (nical) [inline wgpu-hal/src/lib.rs]:
```suggestion
        /// Format can be used as storage with both read and write access.
```

File: wgpu-hal/src/metal/adapter.rs [hunk 5]
```diff
@@ -351,7 +360,7 @@ impl crate::Adapter for super::Adapter {
             usage: crate::TextureUses::COLOR_TARGET
                 | crate::TextureUses::COPY_SRC
                 | crate::TextureUses::COPY_DST
-                | crate::TextureUses::STORAGE_READ
+                //| crate::TextureUses::STORAGE_READ_ONLY
```
Comment (nical) [inline wgpu-hal/src/metal/adapter.rs]:
```suggestion
```

**Review by nical — COMMENTED** [review]
Nice, thanks!

I think that in the case of bgra8unorm-storage, we should not have to request `TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES` (for write-only access, which seems to be the case for at least the bgra8unorm-storage test). Instead, the write-only capability should be added automatically when the `BGRA8UNORM_STORAGE` feature is requested (and granted).

File: wgpu-core/src/device/resource.rs (hunk 3 above)
Comment (atlv24) [reply wgpu-core/src/device/resource.rs:2517]: so the logic here is checking if it has either READ_ONLY or READ_WRITE (intersects, not contains) and if so, then we can do reads on this.

File: wgpu-core/src/device/resource.rs (hunk 2 above)
Comment (atlv24) [reply wgpu-core/src/device/resource.rs:2508]: so the logic here is checking if it has either WRITE_ONLY or READ_WRITE (intersects, not contains) and if so, then we can do write on this.

File: wgpu-types/src/lib.rs [hunk 6]
```diff
@@ -3533,6 +3533,10 @@ impl TextureFormat {
 
         flags.set(TextureFormatFeatureFlags::FILTERABLE, is_filterable);
         flags.set(TextureFormatFeatureFlags::BLENDABLE, is_blendable);
+        flags.set(
+            TextureFormatFeatureFlags::STORAGE_WRITE_ONLY,
+            allowed_usages.contains(TextureUsages::STORAGE_BINDING),
+        );
```
Comment (nical) [inline wgpu-types/src/lib.rs]: This isn't quite right, looks like my earlier comment was misleading. According to [this table](https://www.w3.org/TR/webgpu/#texture-format-caps) `bgra8Unorm` textures have the `STORAGE_WRITE_ONLY` capability if the "bgra8unorm-storage" feature has been requested.

So the condition for setting this flag should be `*self == Self::Bgra8Unorm && device_features.contains(Features::BGRA8UNORM_STORAGE)
`

File: wgpu-core/src/device/resource.rs (hunk 2 above)
Comment (nical) [reply wgpu-core/src/device/resource.rs:2508]: ah, indeed, my bad!

**Comment by atlv24** [comment]
apologies, i thought i was doing a favor fixing the fact that storage write wasnt being set for other storages. pushed a better fix

**Review by nical — APPROVED** [review]

**Comment by nical** [comment]
Looks like some of the tests where you previously added "TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES" now need "BGRA8UNORM_STORAGE".

**Comment by atlv24** [comment]
@nical I just tested locally and thats not enough to fix it -- zero-init-texture-binding.ron uses Rgba8Unorm not Bgra8Unorm so the feature doesnt apply.

File: wgpu-hal/src/lib.rs [hunk 7]
```diff
@@ -1672,8 +1674,10 @@ bitflags::bitflags! {
         /// A uniform buffer bound in a bind group.
         const UNIFORM = 1 << 6;
         /// A read-only storage buffer used in a bind group.
-        const STORAGE_READ = 1 << 7;
-        /// A read-write or write-only buffer used in a bind group.
+        const STORAGE_READ_ONLY = 1 << 7;
+        /// A write-only storage buffer used in a bind group.
+        const STORAGE_WRITE_ONLY = 1 << 8;
```
Comment (teoxoy) [inline wgpu-hal/src/lib.rs:1679]: This occupies the same bit as `STORAGE_READ_WRITE`.

File: wgpu-hal/src/lib.rs [hunk 8]
```diff
@@ -1716,17 +1720,19 @@ bitflags::bitflags! {
         /// Read-write depth stencil usage
         const DEPTH_STENCIL_WRITE = 1 << 7;
         /// Read-only storage buffer usage. Corresponds to a UAV in d3d, so is exclusive, despite being read only.
-        const STORAGE_READ = 1 << 8;
-        /// Read-write or write-only storage buffer usage.
+        const STORAGE_READ_ONLY = 1 << 8;
+        /// Write-only storage buffer usage.
+        const STORAGE_WRITE_ONLY = 1 << 9;
```
Comment (teoxoy) [inline wgpu-hal/src/lib.rs:1725]: This occupies the same bit as `STORAGE_READ_WRITE`.

File: wgpu-hal/src/metal/adapter.rs [hunk 9]
```diff
@@ -111,7 +111,9 @@ impl crate::Adapter for super::Adapter {
 
         // Metal defined pixel format capabilities
         let all_caps = Tfc::SAMPLED_LINEAR
-            | Tfc::STORAGE_WRITE
+            | Tfc::STORAGE_READ_ONLY
+            | Tfc::STORAGE_WRITE_ONLY
+            | Tfc::STORAGE_READ_WRITE
```
Comment (teoxoy) [inline wgpu-hal/src/metal/adapter.rs:116]: `all_caps` should not include read-write usage; it should be based on the read-write tier.

File: wgpu-types/src/lib.rs [hunk 10]
```diff
@@ -3405,10 +3411,13 @@ impl TextureFormat {
         } else {
             basic
         };
-        let bgra8unorm = if device_features.contains(Features::BGRA8UNORM_STORAGE) {
-            attachment | TextureUsages::STORAGE_BINDING
+        let (bgra8unorm_f, bgra8unorm) = if device_features.contains(Features::BGRA8UNORM_STORAGE) {
+            (
+                msaa_resolve | TextureFormatFeatureFlags::STORAGE_WRITE_ONLY,
```
Comment (teoxoy) [inline wgpu-types/src/lib.rs:3416]: With the changes in this PR, all the formats below that support `TextureUsages::STORAGE_BINDING` need at least one of the `TextureFormatFeatureFlags::STORAGE_*` flags to be enabled to be used as storage textures.

`wgpu::Features::TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES` bypasses `guaranteed_format_features` and that's why it gets the ray tracing example and one of the player tests passing.

With this PR it is effectively impossible to use/bind a storage texture (unless you enable `TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES`; which should not be required).

Running `cargo run --bin wgpu-examples storage_texture` results in:

> wgpu error: Validation Error
> 
> Caused by:
>   In Device::create_bind_group
>     The adapter does not support write access for storage textures of format Rgba8Unorm

**Pull request merged**